plugin/src/main/java/net/codingarea/challenges/plugin/challenges/implementation/challenge/LevelEffectChallenge.java

```diff
@@ -0,0 +1,231 @@
+     * because of the random spawning mechanic at the world spawn.
+     */
+    @EventHandler(priority = EventPriority.HIGH)
+    public void onRespawn(@Nonnull PlayerRespawnEvent event) {
+        if (!shouldExecuteEffect()) return;
+        if (ignorePlayer(event.getPlayer())) return;
+        Bukkit.getScheduler().runTaskLater(plugin, this::updateEffects, 1);
+    }
+
+    /**
+     * Execute level change event when dying instead of respawning like spigot does it
+     */
+    @EventHandler(priority = EventPriority.HIGH)
+    public void onDeath(@Nonnull PlayerDeathEvent event) {
+        if (!shouldExecuteEffect()) return;
+        if (ignorePlayer(event.getEntity())) return;
+        PlayerLevelChangeEvent lvlEvent = new PlayerLevelChangeEvent(
+                event.getEntity(), event.getEntity().getLevel(), 0);
+        event.getEntity().setLevel(0);
+        onLevelChange(lvlEvent);
+    }
+
+    @Override
+    public void loadGameState(@NotNull Document document) {
+        super.loadGameState(document);
+        Logger.error("LoadGameState");
+        List<Integer> loadedEffects = document.getIntegerList("loadedEffects");
+        Logger.error("loadedEffects", loadedEffects.toString());
+        List<PotionEffectType> allEffects = Arrays.asList(PotionEffectType.values());
+        amplifierList = document.getIntegerList("amplifier");
+        if(loadedEffects.isEmpty()){
+            effectTypes = allEffects;
+            Collections.shuffle(effectTypes);
+        }
+        else{
+            Map<Integer, PotionEffectType> potionEffectTypeMap = new HashMap<>();
+            for(PotionEffectType effect: allEffects){
+                potionEffectTypeMap.put(effect.getId(), effect);
+            }
+            effectTypes = loadedEffects.stream().map(potionEffectTypeMap::get).collect(Collectors.toList());
+        }
+        if (isEnabled()) {
+            updateEffects();
+        }
+    }
+
+    @Override
+    public void writeGameState(@NotNull Document document) {
+        super.writeGameState(document);
+        List<Integer> effects = effectTypes.stream().map(PotionEffectType::getId).collect(Collectors.toList());
+        Logger.error("writeGameState", effects.toString());
+        document.set("loadedEffects", effects);
+        document.set("amplifier", amplifierList);
+    }
+
+
+    @Nonnull
+    @Override
+    public ItemBuilder createSettingsItem() {
+        int amount = 1;
+        if(isEnabled()){
+            amount = getValue();
+        }
+        if(amount == getMaxValue()){
+           return super.createSettingsItem().setType(Material.COMMAND_BLOCK).setAmount(1);
+        }
+        return super.createSettingsItem().amount(isEnabled() ? getValue() : 1);
+    }
+
+    @Override
+    public void playValueChangeTitle() {
+        ChallengeHelper.playChallengeHeartsValueChangeTitle(this, getValue() * 100);
+    }
+
+
+    @Override
+    protected void onValueChange() {
+        amplifierList = new ArrayList<>();
```

Comment:
Why do you initialize the array?

Reply:
For the sake of simplicity and because this PR is closed anyway I will answer to the points here

First of, **thank you so much for the time** you put into reviewing this. Like I mentioned, these are my first hours of working with a minecraft plugin and frankly put, this one is quite complex. 
This PR was not intended to be an important and valuable addition, it was more there to help me get feedback so that I can maybe create something valuable in the future.
The point was not to make sure I don't merge a main class that I accidentally commited. It was to see if I had the correct structure

The gamestate is very important, because the challenge is based around the idea that the effects remain the same. So if you want to keep an effect you need to find ways to level down as well. If this wouldn't be the case you could simply level up -> effect -> ambos -> level down -> new effect and grind yourself through effects, which is not a fun way to play the game.
Maybe only leveling up should give you a new effect
I wanted to make these things configureable, but it was really difficult to get even this Range for the amplifier to work.

**How would I add multiple parameters into the challenge. Like "save effects" "amplifier from x to y" "only when levelling up"?**